### PR TITLE
[Merged by Bors] - feat(Ideal/IsPrincipalPowQuotient): R/I equiv I^n/I^(n+1)

### PR DESCRIPTION
**Pull request by pechersky**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+variable {R : Type*} [CommRing R] [IsDomain R] {I : Ideal R}
+
+/-- For a principal ideal `I`, `R ⧸ I ≃ₗ[R] I ^ n ⧸ I ^ (n + 1)`. To convert into a form
+that uses the ideal of `R ⧸ I ^ (n + 1)`, compose with
+`Ideal.pow_quot_pow_succ_equiv_map_mk_pow_succ_pow`. -/
+noncomputable
+def quot_equiv_pow_quot_pow_succ (h : I.IsPrincipal) (h': I ≠ ⊥) (n : ℕ) :
+    (R ⧸ I) ≃ₗ[R] (I ^ n : Ideal R) ⧸ (I • ⊤ : Submodule R (I ^ n : Ideal R)) := by
+  let f : (I ^ n : Ideal R) →ₗ[R] (I ^ n : Ideal R) ⧸ (I • ⊤ : Submodule R (I ^ n : Ideal R)) :=
+      Submodule.mkQ _
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean:43]:
```suggestion
    Submodule.mkQ _
```

File: Mathlib/RingTheory/Ideal/QuotientOperations.lean [hunk 2]
```diff
@@ -932,3 +932,47 @@ theorem quotQuotEquivQuotOfLE_symm_comp_mkₐ (h : I ≤ J) :
...
+    simp
+
+/-- I ^ n ⧸ I ^ (n + 1) is equivalent as a quotient of I ^ n or as an ideal of
+R ⧸ I ^ (n + 1). Supplied as a plain equiv to bypass typeclass synthesis issues on complex
+`Module` goals. -/
+noncomputable
+def pow_quot_pow_succ_equiv_map_mk_pow_succ_pow_equiv :
+    ((I ^ n : Ideal R) ⧸ (I • ⊤ : Submodule R (I ^ n : Ideal R))) ≃
+    Ideal.map (Ideal.Quotient.mk (I ^ (n + 1))) (I ^ n) :=
+pow_quot_pow_succ_equiv_map_mk_pow_succ_pow I n
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Ideal/QuotientOperations.lean:974]:
```suggestion
  pow_quot_pow_succ_equiv_map_mk_pow_succ_pow I n
```

File: Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+section IsPrincipal
+
+variable {R : Type*} [CommRing R] [IsDomain R] {I : Ideal R}
+
+/-- For a principal ideal `I`, `R ⧸ I ≃ₗ[R] I ^ n ⧸ I ^ (n + 1)`. To convert into a form
+that uses the ideal of `R ⧸ I ^ (n + 1)`, compose with
+`Ideal.pow_quot_pow_succ_equiv_map_mk_pow_succ_pow`. -/
+noncomputable
+def quot_equiv_pow_quot_pow_succ (h : I.IsPrincipal) (h': I ≠ ⊥) (n : ℕ) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean:40]: Hmm, actually, shouldn't this use camelCase?

Comment (pechersky) [reply Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean]: changed

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

File: Mathlib/RingTheory/Ideal/QuotientOperations.lean [hunk 4]
```diff
@@ -932,3 +932,47 @@ theorem quotQuotEquivQuotOfLE_symm_comp_mkₐ (h : I ≤ J) :
...
+namespace Ideal
+
+section PowQuot
+
+variable {R : Type*} [CommRing R] (I : Ideal R) (n : ℕ)
+
+/-- I ^ n ⧸ I ^ (n + 1) is equivalent as a quotient of I ^ n or as an ideal of
+R ⧸ I ^ (n + 1). -/
+noncomputable
+def pow_quot_pow_succ_equiv_map_mk_pow_succ_pow :
```
Comment (jcommelin) [inline Mathlib/RingTheory/Ideal/QuotientOperations.lean:945]: Should this be camel-cased?

File: Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace Ideal
+
+section IsPrincipal
+
+variable {R : Type*} [CommRing R] [IsDomain R] {I : Ideal R}
+
+/-- For a principal ideal `I`, `R ⧸ I ≃ₗ[R] I ^ n ⧸ I ^ (n + 1)`. To convert into a form
+that uses the ideal of `R ⧸ I ^ (n + 1)`, compose with
+`Ideal.pow_quot_pow_succ_equiv_map_mk_pow_succ_pow`. -/
```
Comment (jcommelin) [inline Mathlib/RingTheory/Ideal/IsPrincipalPowQuotient.lean:38]: Please update this name accordingly, if you change it in the other file.

File: Mathlib/RingTheory/Ideal/QuotientOperations.lean [hunk 6]
```diff
@@ -932,3 +932,47 @@ theorem quotQuotEquivQuotOfLE_symm_comp_mkₐ (h : I ≤ J) :
...
 end DoubleQuot
+
+namespace Ideal
+
+section PowQuot
+
+variable {R : Type*} [CommRing R] (I : Ideal R) (n : ℕ)
+
+/-- I ^ n ⧸ I ^ (n + 1) is equivalent as a quotient of I ^ n or as an ideal of
+R ⧸ I ^ (n + 1). -/
```
Comment (jcommelin) [inline Mathlib/RingTheory/Ideal/QuotientOperations.lean:943]: I had a bit of trouble parsing the docstring upon first reading. Not sure if the following is an improvement though
```suggestion
/-- `I ^ n ⧸ I ^ (n + 1)` can be viewed as a quotient module and as ideal of `R ⧸ I ^ (n + 1)`.
This definition gives the `R`-linear equivalence between the two. -/
```

File: Mathlib/RingTheory/Ideal/QuotientOperations.lean [hunk 7]
```diff
@@ -932,3 +932,47 @@ theorem quotQuotEquivQuotOfLE_symm_comp_mkₐ (h : I ≤ J) :
...
+    simp [Ideal.Quotient.eq, Submodule.Quotient.eq, Submodule.mem_smul_top_iff, pow_succ']
+  · intro ⟨x, hx⟩
+    rw [Ideal.mem_map_iff_of_surjective _ Ideal.Quotient.mk_surjective] at hx
+    obtain ⟨y, hy, rfl⟩ := hx
+    refine ⟨Submodule.Quotient.mk ⟨y, hy⟩, ?_⟩
+    simp
+
+/-- I ^ n ⧸ I ^ (n + 1) is equivalent as a quotient of I ^ n or as an ideal of
+R ⧸ I ^ (n + 1). Supplied as a plain equiv to bypass typeclass synthesis issues on complex
+`Module` goals. -/
```
Comment (jcommelin) [inline Mathlib/RingTheory/Ideal/QuotientOperations.lean:969]: Idem

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**